gateway-service-impl/src/main/java/org/hypertrace/gateway/service/entity/query/ExecutionTreeBuilder.java

```diff
@@ -357,4 +356,54 @@ private QueryNode createQsDataFetcherNodeWithLimitAndOffset(EntitiesRequest enti
   private QueryNode createPaginateOnlyNode(QueryNode queryNode, EntitiesRequest entitiesRequest) {
     return new PaginateOnlyNode(queryNode, entitiesRequest.getLimit(), entitiesRequest.getOffset());
   }
+
+  private Optional<String> getSingleSourceForAllAttributes() {
+    // All expressions' attributes from the same source. Will only need one downstream query.
+    Optional<String> singleSourceForAllAttributes =
+        ExecutionTreeUtils.getSingleSourceForAllAttributes(executionContext);
+    EntitiesRequest entitiesRequest = executionContext.getEntitiesRequest();
+
+    if (singleSourceForAllAttributes.isEmpty()) {
+      return Optional.empty();
+    }
+
+    /**
+     * Entities queries usually have an inherent time filter, via {@link
+     * EntitiesRequest#getStartTimeMillis()} and {@link EntitiesRequest#getEndTimeMillis()}.
+     *
+     * <p>The time filter is usually applied on QS, apart from few entity types, which have a
+     * timestamp column specified through timestamp config {@link TimestampConfigs}
+     *
+     * <p>Single source for all attributes doesn't take care of the inherent time filter for the
+     * single source
+     *
+     * <p>If the single source is {@link EDS} and the timestamp column doesn't exist for the
+     * corresponding entity type, the query will be executed on {@link EDS} source, without the
+     * inherent filter, leading to erroneous outputs
+     *
+     * <p>Hence, default to non single source, if
+     *
+     * <ul>
+     *   <li>single source is {@link EDS}
+     *   <li>inherent time filter is valid
+     *   <li>only live entities are requested (since non live entities request ignores the inherent
+     *       time filter anyways)
+     *   <li>timestamp column doesn't exist for the corresponding entity type
+     * </ul>
+     */
+    String singleSource = singleSourceForAllAttributes.get();
+    if (EDS.name().equals(singleSource)
+        && isValidTimeRange(
```

Comment:
Do we support queries without a time range today? Doesn't seem like the spot to be doing validation

Reply:
Can get rid of the validation, but seemed like a valid validation, since it's moved under utils now

Reply:
it's not that it's not valid - it's just not the purpose of this method - single responsibility!